src/controller.coffee

```diff
@@ -424,25 +429,108 @@ Editor::redrawMain = (opts = {}) ->
       @mainCtx.save()
       opts.boundingRectangle.clip @mainCtx
 
-    # Draw the new tree on the main context
-    layoutResult = @view.getViewNodeFor(@tree).layout 0, @nubbyHeight
-    @view.getViewNodeFor(@tree).draw @mainCtx, opts.boundingRectangle ? new @draw.Rectangle(
+    rect = opts.boundingRectangle ? new @draw.Rectangle(
       @scrollOffsets.main.x,
       @scrollOffsets.main.y,
       @mainCanvas.width,
       @mainCanvas.height
-    ), {
+    )
+    options = {
       grayscale: false
       selected: false
       noText: (opts.noText ? false)
     }
 
-    # Draw the cursor (if exists, and is inserted)
-    @redrawCursors(); @redrawHighlights()
+    # Draw the new tree on the main context
+    layoutResult = @view.getViewNodeFor(@tree).layout 0, @nubbyHeight
+    @view.getViewNodeFor(@tree).draw @mainCtx, rect, options
+
+    @mainCtx.globalAlpha *= FLOATING_BLOCK_ALPHA
+
+    # Draw floating blocks
+    startWidth = @mainCtx.measureText(@mode.startComment).width
+    endWidth = @mainCtx.measureText(@mode.endComment).width
+    for record in @floatingBlocks
+      blockView = @view.getViewNodeFor record.block
+      blockView.layout record.position.x, record.position.y
+
+      rectangle = new @view.draw.Rectangle(); rectangle.copy(blockView.totalBounds)
+      rectangle.x -= GRAY_BLOCK_MARGIN; rectangle.y -= GRAY_BLOCK_MARGIN
+      rectangle.width += 2 * GRAY_BLOCK_MARGIN; rectangle.height += 2 * GRAY_BLOCK_MARGIN
+
+      bottomTextPosition = blockView.totalBounds.bottom() - blockView.distanceToBase[blockView.lineLength - 1].below - @fontSize
```

Comment:
this is a lot of complexity to put inline in the main function.  Can you factor out drawing each floating block into its own function, perhaps?